lib/tasks/sync_users.rake

```diff
@@ -0,0 +1,187 @@
+    # if barcode is set, then only checks that barcode for Sierra-MLN mismatch
+    args.with_defaults(:file_name => nil, :start=> 0, :limit => 0, :barcode => nil)
+
+    csv_start = args.start.to_i
+    csv_limit = args.limit.to_i
+
+    csv_barcode = args.barcode
+
+    # Example:  'db/migrate/20181128_sierra_mln_user_accounts.csv'
+    csv_path = args.file_name
+
+    # \u03EA is a unicode character that will never happen in our Sierra output file.
+    # we need this, because the Sierra output has some junk data that's making the CSV gem trip up.
+    # however, this means that we'll now need to strip the double quotes from around the ingested field values ourselves.
+    CSV.foreach(csv_path, col_sep: "|", quote_char: "\u03EA") do |line|
+      # don't process the header row
+      next if $. == 1
+
+      # if we're only reading a few lines of the csv, then see if we can stop, before reading any further
+      next if (csv_start > 0 && $. < csv_start)
+      break if (csv_limit > 0 && $. >= (csv_start + csv_limit))
+
+      # "P BARCODE"|"EMAIL ADDR"|"PIN"|"EXP DATE"|"PCODE3"|"P TYPE"|"TOT CHKOUT"|"HOME LIBR"|"MBLOCK"|"PCODE4"|"PATRN NAME"|"ADDRESS"|"CREATED(PATRON)"|"UPDATED(PATRON)"
+      # Example:  "23333106701234"|"name@gmail.com"|"Tn/jC7sHXQpfw"|"10-01-2018"|"1"|"153"|"4"|"ea   "|"-"|"895"|"LASTNAME, FIRSTNAME"|"123 MAIN ST, NY 10001"|"05-20-2009 15:23"|"07-02-2018"
+      (barcode, email, pin, expiration, pcode3, ptype, total_checkouts, home_library, manual_block, pcode4, name, address, sierra_created, sierra_updated) = line
+
+      # inore bad data, and move on to the next line
+      next if (barcode.nil?)
+
+      # clean off any leading/trailing spaces and double quotes
+      barcode = barcode.strip.downcase
+      barcode = barcode.chomp('"').reverse.chomp('"').reverse
+      sierra_created = sierra_created.chomp('"').reverse.chomp('"').reverse
+
+      # if a barcode was passed, we're looking for that barcode, and should stop processing once we find it.
+      if (!csv_barcode.nil? && csv_barcode == barcode)
+        csv_start = $.
+        csv_limit = 1
+      end
+
+      # is there a user in the mln database that matches this user from Sierra?
+      mln_user = User.find_by_barcode(barcode)
+      if !mln_user.nil?
+        # Yes, there is.  Now see if the user was created in MLN over the last month, while being old in Sierra.
+
+        # was the user created after a month ago in mln database?
+        if (mln_user.created_at > 1.month.ago)
+          begin
+            date_sierra_created = DateTime.strptime(sierra_created, '%m-%d-%Y %H:%M')
+
+            # was the user created before a month ago in Sierra?
+            if (date_sierra_created < 1.month.ago)
+              puts "manually repaired user=#{barcode}, mln_user.id=#{mln_user.id}, mln_user.created_at=#{mln_user.created_at}, mln_user.updated_at=#{mln_user.updated_at}, sierra_created=#{sierra_created}, sierra_updated=#{sierra_updated}"
```

Comment:
Do you need to do anything to the user record in the database, or is printing the info sufficient?

Reply:
not at this time, I'm making the decision to not complete the functionality requested, because auto-inserting user records is slightly risky, and the results of the task I did run show a too-low number of records to be fixed to justify the risks.